### ISSUE TITLE
Lookbook Listening 

### ISSUE THREAD

**Issue by jonsgreen** [ViewComponent/lookbook#203]
I am having trouble getting the listener to catch preview changes. I regret bothering you with this here as an issue since I am sure it is something I am doing but if you have any suggestions about the best way to debug this please let me know. Thanks so much.

**Comment by allmarkedup** [comment]
Het @jonsgreen - sorry to hear you are having troubles and no worries about opening the issue!

Before diving any deeper, if you are using Puma for your webserver can I ask if you've seen the [troubleshooting note here](https://lookbook.build/guide/troubleshooting/#-forking-servers---changes-not-showing-up)? It's a pretty common cause of this issue so just want to check you've seen that first.

If that doesn't apply to you (or the suggested workaround there doesn't fix things!) just let me know and we can figure out some next steps to diagnose the issue.

**Comment by jonsgreen** [comment]
@allmarkedup I had seen that and checked that my dev puma server is running in single mode. I am running my server using [overmind](https://github.com/DarthSim/overmind) instead of foreman; not sure if that could be a factor?

**Comment by allmarkedup** [comment]
@jonsgreen ok great, always good to rule out the most likely culprits first :-)

Overmind could indeed be a factor - I have to admit I've never used it or looked at it in any detail but I'll try to run a quick test with it later today and see if I can recreate the issue.

In the meantime, would you mind sharing some debug data so I can eyeball it for anything unusual? In the top right of the Lookbook UI is a little ? icon in a circle which opens a dropdown debug menu:

<img width="211" alt="194727253-ee961698-1e75-435a-ad5f-4e7a00bd4f0e" src="https://user-images.githubusercontent.com/126726/198032154-956785f7-f353-40c7-9873-b7d8273941bc.png">

Could you click the 'copy debug data' link and paste in the output here for me to look at? It shouldn't contain any sensitive information (apart from potentially the file paths) but obviously redact anything you feel you need to.

I'll let you know if I turn anything up once I've given overmind a spin.

**Comment by jonsgreen** [comment]
Hey @allmarkedup, here is the debug data you asked for. Sorry about the delay: 

```ruby
{
  "version": "1.2.1",
  "env": "development",
  "config": {
    "project_name": "GTO Design System LookBook",
    "log_level": 2,
    "log_use_rails_logger": true,
    "auto_refresh": true,
    "components_path": "app/components",
    "page_controller": "Lookbook::PageController",
    "page_route": "pages",
    "page_paths": [
      "test/components/docs"
    ],
    "page_options": {
    },
    "markdown_options": {
      "tables": true,
      "fenced_code_blocks": true,
      "disable_indented_code_blocks": true,
      "strikethrough": true,
      "highlight": true,
      "with_toc_data": true,
      "lax_spacing": true
    },
    "sort_examples": false,
    "preview_paths": [
      "spec/components/previews",
      "spec/components/previews",
      "/redacted/spec/components/previews"
    ],
    "preview_display_options": {
    },
    "preview_disable_action_view_annotations": true,
    "preview_params_options_eval": false,
    "listen": true,
    "listen_paths": [
      "spec/components/previews",
      "spec/components/previews",
      "/redacted/spec/components/previews",
      "app/components"
    ],
    "listen_extensions": [
      "rb",
      "html.*"
    ],
    "listen_use_polling": false,
    "ui_theme": "indigo",
    "ui_theme_overrides": {
    },
    "ui_favicon": true,
    "debug_menu": true,
    "experimental_features": false,
    "preview_controller": "ViewComponentsController",
    "panels": {
      "main": [
        {
          "label": "Preview",
          "hotkey": "v",
          "disabled": false,
          "show": true,
          "copy": null,
          "locals": {
          },
          "partial": "lookbook/previews/panels/preview",
          "name": "preview",
          "group": "main"
        },
        {
          "label": "HTML",
          "hotkey": "h",
          "disabled": false,
          "show": true,
          "copy": null,
          "locals": {
          },
          "partial": "lookbook/previews/panels/output",
          "name": "output",
          "group": "main"
        }
      ],
      "drawer": [
        {
          "label": "Source",
          "hotkey": "s",
          "disabled": false,
          "show": true,
          "copy": "#<Proc:0x0000000114c81d00 (eval):1 (lambda)>",
          "locals": {
          },
          "partial": "lookbook/previews/panels/source",
          "name": "source",
          "group": "drawer"
        },
        {
          "label": "Notes",
          "hotkey": "n",
          "disabled": "#<Proc:0x0000000114c81800 (eval):1 (lambda)>",
          "show": true,
          "copy": null,
          "locals": {
          },
          "partial": "lookbook/previews/panels/notes",
          "name": "notes",
          "group": "drawer"
        },
        {
          "label": "Params",
          "hotkey": "p",
          "disabled": "#<Proc:0x0000000114c81440 (eval):1 (lambda)>",
          "show": true,
          "copy": null,
          "locals": {
          },
          "partial": "lookbook/previews/panels/params",
          "name": "params",
          "group": "drawer"
        }
      ]
    },
    "inputs": {
      "select": {
        "name": "select",
        "partial": "lookbook/previews/inputs/select",
        "opts": {
          "opts": {
          }
        }
      },
      "textarea": {
        "name": "textarea",
        "partial": "lookbook/previews/inputs/textarea",
        "opts": {
          "opts": {
          }
        }
      },
      "toggle": {
        "name": "toggle",
        "partial": "lookbook/previews/inputs/toggle",
        "opts": {
          "opts": {
          }
        }
      },
      "color": {
        "name": "color",
        "partial": "lookbook/previews/inputs/color",
        "opts": {
          "opts": {
          }
        }
      },
      "range": {
        "name": "range",
        "partial": "lookbook/previews/inputs/range",
        "opts": {
          "opts": {
          }
        }
      },
      "text": {
        "name": "text",
        "partial": "lookbook/previews/inputs/text",
        "opts": {
          "opts": {
          }
        }
      },
      "email": {
        "name": "email",
        "partial": "lookbook/previews/inputs/text",
        "opts": {
          "opts": {
          }
        }
      },
      "number": {
        "name": "number",
        "partial": "lookbook/previews/inputs/text",
        "opts": {
          "opts": {
          }
        }
      },
      "tel": {
        "name": "tel",
        "partial": "lookbook/previews/inputs/text",
        "opts": {
          "opts": {
          }
        }
      },
      "url": {
        "name": "url",
        "partial": "lookbook/previews/inputs/text",
        "opts": {
          "opts": {
          }
        }
      },
      "date": {
        "name": "date",
        "partial": "lookbook/previews/inputs/text",
        "opts": {
          "opts": {
          }
        }
      },
      "datetime_local": {
        "name": "datetime_local",
        "partial": "lookbook/previews/inputs/text",
        "opts": {
          "opts": {
          }
        }
      }
    },
    "tags": {
      "hidden": {
        "name": "hidden",
        "label": "Hidden status",
        "yard_parser": null,
        "opts": {
        }
      },
      "label": {
        "name": "label",
        "label": "Label",
        "yard_parser": null,
        "opts": {
        }
      },
      "display": {
        "name": "display",
        "label": "Display",
        "yard_parser": null,
        "opts": {
        }
      },
      "position": {
        "name": "position",
        "label": "Position",
        "yard_parser": null,
        "opts": {
        }
      },
      "id": {
        "name": "id",
        "label": "Id",
        "yard_parser": null,
        "opts": {
        }
      },
      "component": {
        "name": "component",
        "label": "Component",
        "yard_parser": null,
        "opts": {
        }
      },
      "param": {
        "name": "param",
        "label": "Param",
        "yard_parser": "with_name",
        "opts": {
        }
      }
    }
  }
}
```

**Comment by allmarkedup** [comment]
Thanks for that @jonsgreen - however it all looks good to me, nothing obviously problematic there at all 😕

I've also just tried using Overmind to run an app with Lookbook installed and listening for changes seems to work fine, even when running multiple instances of each process. So i don't think it can be that.

One thing to try - can you set the value of the `listen_use_polling` [config option](https://lookbook.build/api/config/#system:~:text=%22css%22%5D-,listen_use_polling,-Type%3A%20Boolean) to `true` and see if that makes any difference?

```rb
config.lookbook.listen_use_polling = true
```

Sorry this is not a great experience for you right now!

**Comment by allmarkedup** [comment]
Just linking these discussions here: https://github.com/ViewComponent/lookbook/discussions/148 as they seem like they might be to do with the same issue (whatever it is!)

**Comment by jonsgreen** [comment]
@allmarkedup turning on `listen_use_polling` did not seem to help. I think you know this already but it is just preview changes that are not getting updated; changes to the actual component are updated in Lookbook. 

It does seem like the user in the other issue is having the same problem.

**Comment by mdrbohlav** [comment]
I had exactly the same issue but found out that even if the `WEB_CONCURRENCY` is set to `1` the Puma webserver (at least version 6.0.0) does run in a clustered mode. Try setting it to `0` instead, that fixed it for me.

**Comment by jonsgreen** [comment]
@mdrbohlav thanks for the suggestion. Unfortunately that did not seem to fix it for me.

**Comment by allmarkedup** [comment]
@jonsgreen really sorry for the slow reply, I've had a bit of a hectic week.

I hadn't actually picked up on the fact that it was only changes to the previews (not the component files themselves) that were the problem. That definitely helps narrow down the list of things that might be causing a problem here as it means that the file change listening isn't failing completely - it's just not registering changes in your previews directory for some reason.

If you add your preview directory (it looks like `spec/components/previews` from your debug data dump) directly to the `lookbook.listen_paths` config option does that have any effect? 

```rb
config.lookbook.listen_paths = ["spec/components/previews"]
```

Is the repo you are having trouble with open-source? If so can you point me to it so I can clone it and see if I can replicate the issue within it?

**Comment by allmarkedup** [comment]
@mdrbohlav Thanks for the heads up about the puma WEB_CONCURRENCY value - I'll update the troubleshooting section in the docs to reflect that 👍

**Comment by jonsgreen** [comment]
@allmarkedup thanks for your persistence in trying to help me figure this out. 

I already had that configuration. This is what I have in config/application.rb:

```ruby
  config.view_component.generate_sidecar = true
  config.view_component.preview_paths += [Rails.root.join('spec/components/previews')]
  config.lookbook.project_name = 'GTO Design System LookBook'
  config.lookbook.preview_paths = ['spec/components/previews']
  config.view_component.default_preview_layout = 'lookbook'
```
I am not sure exactly what the behavior should be like but I should explain that I never get hot code reloading if I make changes to either the components or the previews. However, if I make changes to either the component code or its corresponding erb template and refresh Lookbook then it picks up what has changed. I also do get updates if I change the preview erb template and refresh the page. However, any changes that I make to the preview ruby file are not picked up when I refresh the preview in Lookbook. 

Unfortunately, this is a private repository for a client so I cannot share. I suppose I could try to replicate the issue in a vanilla app though it would take some work to get that all setup and not sure I am up for it right now.

**Comment by allmarkedup** [comment]
@jonsgreen no worries, happy to try and help although I'm afraid i'm running out of ideas here!

I've tried running the [demo app](https://github.com/ViewComponent/lookbook-demo) using Overmind and with the preview directory changed to `spec/components/preview` and using the config settings you provided above and unfortunately i cannot replicate the issues you are seeing.

There must be some other config/local setup mismatch going on here but without anything else to go on I'm afraid I'm a bit stumped 😕

Totally understand you not having a lot of time to put into this but if you do have a chance to run the demo app locally and do let me know if you encounter the same issues.

FWIW I've just released a few updates, some of which include some refactoring of the websocket/listening code. I don't anticipate them having any effect on your problems but it's probably worth updating just in case!
